File: dat-pypi/tests/test_cms_fixture.py
import json
from pathlib import Path
from unittest.mock import patch

import pytest

from saro_dat import DatCmsManager, DatManager
from saro_dat import error as E
from saro_dat.error import DatError
from saro_dat.dat_cms_manager import _SameOriginRedirect


FIXTURE = json.loads((Path(__file__).parent / "fixtures" / "cms_v1_state_transitions.json").read_text())


class Response:
    def __init__(self, status, body):
        self.status = status
        self._body = body

    def read(self):
        return self._body

    def __enter__(self):
        return self

    def __exit__(self, *_):
        return False


class Opener:
    def __init__(self, response=None, error=None):
        self.response = response
        self.error = error
        self.timeout = None

    def open(self, *_args, **kwargs):
        self.timeout = kwargs.get("timeout")
        if self.error:
            raise self.error
        return self.response


def body(case):
    result = b""
    for kind, value in case["input"].get("body", []):
        if kind == "ascii":
            result += value.encode("ascii")
        elif kind == "hex":
            result += bytes.fromhex(value)
        else:
            result += FIXTURE["certificates"][value]["wire_ascii"].encode("ascii")
    return result


def make_manager(state):
    manager = DatCmsManager.__new__(DatCmsManager)
    manager._uri = "http://fixture.invalid/v1/certs"
    manager._token = ""
    manager._manager = DatManager()
    manager._version = int(FIXTURE["states"][state]["version"])
    manager._sync_lock = __import__("threading").Lock()
    manager._connect_timeout_seconds = 0
    manager._sync_timeout_seconds = 0
    for key in FIXTURE["states"][state]["certificates"]:
        manager._manager.imports(FIXTURE["certificates"][key]["wire_ascii"], False)
    return manager


def expected(case):
    return case.get("expect") or case["expect_by_profile"]["unsigned_u64"]


@pytest.mark.parametrize("case", FIXTURE["cases"], ids=lambda case: case["id"])
def test_g0_cms_state_transitions(case):
    cms = make_manager(case["initial"])
    want = expected(case)
    if case["input"]["kind"] == "transport":
        opener = Opener(error=__import__("urllib.error").error.URLError("transport"))
    else:
        opener = Opener(Response(case["input"]["status"], body(case)))
    with patch("urllib.request.build_opener", return_value=opener):
        if want["error"]:
            with pytest.raises(DatError) as raised:
                cms.sync_or_raise()
            assert raised.value.code == want["error"].split("(")[0]
            assert raised.value.retry.value == want["retry"]
        else:
            cms.sync_or_raise()
    state = FIXTURE["states"][want["state"]]
    assert cms._version == int(state["version"])
    assert [cert.cid for cert in cms._manager._certificates] == [
        int(FIXTURE["certificates"][key]["cid"], 16) for key in state["certificates"]
    ]


def test_same_origin_redirect_rejects_other_origin():
    handler = _SameOriginRedirect(("https", "cms.example"))
    with pytest.raises(__import__("urllib.error").error.URLError):
        handler.redirect_request(None, None, 302, "found", {}, "https://other.example/v1/certs")


def test_same_origin_redirect_is_allowed():
    handler = _SameOriginRedirect(("https", "cms.example"))
    request = __import__("urllib.request").request.Request("https://cms.example/v1/certs")
    assert handler.redirect_request(request, None, 302, "found", {}, "https://cms.example/next")

File: dat-ruby/test/test_cms_fixture.rb
# frozen_string_literal: true

require 'json'
require_relative './test_helper'

class TestCmsFixture < Minitest::Test
  FIXTURE = JSON.parse(File.read(File.expand_path('fixtures/cms_v1_state_transitions.json', __dir__)))

  Response = Struct.new(:code, :body)
  Http = Struct.new(:response) do
    def request(*)
      response
    end
  end

  def body_for(case_data)
    case_data.fetch('input').fetch('body', []).map do |kind, value|
      case kind
      when 'ascii' then value.b
      when 'hex' then [value].pack('H*')
      else FIXTURE.fetch('certificates').fetch(value).fetch('wire_ascii').b
      end
    end.join
  end

  def make_manager(state_name)
    state = FIXTURE.fetch('states').fetch(state_name)
    manager = Saro::Dat::DatManager.new
    state.fetch('certificates').each do |key|
      manager.imports(FIXTURE.fetch('certificates').fetch(key).fetch('wire_ascii'), clear: false)
    end
    cms = Saro::Dat::DatCmsManager.allocate
    cms.instance_variable_set(:@uri, 'http://fixture.invalid/v1/certs')
    cms.instance_variable_set(:@token, '')
    cms.instance_variable_set(:@manager, manager)
    cms.instance_variable_set(:@version, state.fetch('version').to_i)
    cms.instance_variable_set(:@lock, Mutex.new)
    cms.instance_variable_set(:@lifecycle, Mutex.new)
    cms.instance_variable_set(:@connect_timeout_seconds, 0)
    cms.instance_variable_set(:@sync_timeout_seconds, 0)
    cms.instance_variable_set(:@logger, Logger.new(nil))
    cms
  end

  FIXTURE.fetch('cases').each do |case_data|
    define_method("test_g0_#{case_data.fetch('id')}") do
      cms = make_manager(case_data.fetch('initial'))
      want = case_data['expect'] || case_data.fetch('expect_by_profile').fetch('unbounded_decimal')
      input = case_data.fetch('input')
      if input.fetch('kind') == 'transport'
        with_http_start(->(*) { raise IOError, 'transport' }) { assert_fixture(cms, want) }
      else
        response = Response.new(input.fetch('status').to_s, body_for(case_data))
        with_http_start(->(*, &block) { block.call(Http.new(response)) }) { assert_fixture(cms, want) }
      end
      state = FIXTURE.fetch('states').fetch(want.fetch('state'))
      assert_equal state.fetch('version').to_i, cms.version
      assert_equal state.fetch('certificates').map { |key| FIXTURE.fetch('certificates').fetch(key).fetch('wire_ascii') },
                   cms.get_manager.exports.split("\n").reject(&:empty?)
    end
  end

  def assert_fixture(cms, want)
    if want['error']
      error = assert_raises(Saro::Dat::Error) { cms.sync_or_raise }
      assert_equal want.fetch('error').split('(').first, error.code
      assert_equal want.fetch('retry').to_sym, error.retry
    else
      cms.sync_or_raise
    end
  end

  def with_http_start(replacement)
    singleton = Net::HTTP.singleton_class
    original = Net::HTTP.method(:start)
    singleton.send(:define_method, :start, &replacement)
    yield
  ensure
    singleton.send(:define_method, :start, original)
  end
end
